File: WSS_4.6/WSS_4.6/HTTP_ALL.vb
Imports System.Web
Imports System.Net

Imports System.Net.Sockets
Imports System.Text
Imports System.IO
Imports System.Threading
Imports System.Text.Encoding
'Imports WSS.Client

Module http_all
    'Public Declare Sub CopyMemory Lib "Kernel32.dll" Alias "RtlMoveMemory" (ByVal Destination As Any, ByVal Source As Any, ByVal Length As Long)
    Public v_http_conflict_count As ULong

    Public v_protectoin As ULong
    Public chk_Deny_Enable As Boolean = False
    Public Deny_List As New List(Of String)
    Public Users_List As New List(Of String)
    Public sim_1 As Double
    Public sim_2 As Double
    Public sim_3 As Double
    Public send_data As Boolean = False
    Public simul_data As Boolean = False
    Public Wait_Bool As Boolean = False
    Public Clients_LIST As New Clients
    '  Public Servers_LIST As New Servers
    Public count_PAC, count_Byte, count_Byte_S, count_PAC_S, count_Err As Long
    Public count_PAC_REC, count_Byte_REC, count_Byte_S_REC, count_PAC_S_REC, count_Err_REC As Long
    Public Size_For_Send As Integer
    Public Wait_MSeconds As Short
    Public Global_Data_For_Send As String


    Public Function F_ADD_CLIENT(p_url As String, p_Socket As Socket) As Integer
        Dim HostArr As String() = Nothing
        Dim myHost As System.Net.IPHostEntry
        Dim Deny As Boolean = False
        Dim HostName As String = ""
        If Not f_get_allow(p_url) Then
            p_Socket.Close()
            Return -1
        End If
        Try
            HostArr = p_url.Split({":"c})
            myHost = System.Net.Dns.GetHostEntry(HostArr(0)) '.GetHostName
            HostName = myHost.HostName
            If Not f_get_allow(HostName) Then
                p_Socket.Close()
                Return -1
            End If
            Users_List.Add(HostName)
        Catch ex As Exception
            HostName = "no DNSname"
        End Try


        For i As Short = 0 To Clients_LIST.Count - 1
            If Clients_LIST(i).URL = p_url Then
                Return -1
            ElseIf Not Clients_LIST(i).IsActive Then
                Clients_LIST(i).URL = p_url
                Clients_LIST(i).ID = i
                'c_IP = p_IP
                Clients_LIST(i).Connected = Now
                Clients_LIST(i).Stat = 0
                Clients_LIST(i).Socket = p_Socket
                Clients_LIST(i).IsActive = True
                Clients_LIST(i).MachineName = HostName
                'Clients_LIST(i).Conn_Num = i
                Clients_LIST(i).ProcessReceive()
                'Clients_LIST(i).UserName = User
                Return i
            End If
        Next
        Clients_LIST.Add(New Client(p_url, p_Socket))
        Clients_LIST(Clients_LIST.Count - 1).ID = Clients_LIST.Count - 1
        Clients_LIST(Clients_LIST.Count - 1).MachineName = HostName
        Clients_LIST(Clients_LIST.Count - 1).ProcessReceive()
        'Clients_LIST(Clients_LIST.Count - 1).Conn_Num = Clients_LIST.Count - 1
        Return Clients_LIST.Count - 1
    End Function

    Public Function f_get_allow(Str_Deny As String) As Boolean
        'Dim obj As ListBox = System.Windows.Application.Current.MainWindow.FindName("DenyListBox")
        If Not chk_Deny_Enable Then
            Return True
        End If
        Dim v_ret As Boolean = True
        For j As Short = 0 To 1
            Select Case j
                Case 0
                    For i As Short = 0 To Deny_List.Count - 1
                        If InStr(Strings.UCase(Str_Deny), Strings.Right(Deny_List.Item(i).ToString, Strings.Len(Deny_List.Item(i).ToString) - 3)) > 0 And Strings.Left(Deny_List.Item(i).ToString, 3) = "DN=" Then
                            v_ret = False
                        End If
                    Next
                Case 1
                    For i As Short = 0 To Deny_List.Count - 1
                        If InStr(Strings.UCase(Str_Deny), Strings.Right(Deny_List.Item(i).ToString, Strings.Len(Deny_List.Item(i).ToString) - 3)) > 0 And Strings.Left(Deny_List.Item(i).ToString, 3) = "AL=" Then
                            v_ret = True
                        End If
                    Next
            End Select
        Next
        Return v_ret
    End Function
    Public Sub SocketException(ByVal e As Socket, ByVal p_ind As Short)
        Try
            Clients_LIST(p_ind).IsActive = False
            p_errlog("SockExeption ", Strings.Left(Clients_LIST(p_ind).URL & Space(23), 23) & "|" & Strings.Left("Delete_Client" & Space(30), 30) & " | " _
                     & Strings.Left(Clients_LIST(p_ind).UserName & Space(23), 23), "Diconnect")

        Catch ex As Exception

        End Try
        'p_errlog("SocketException", "Delete_Client" & Clients_LIST(p_ind).Connected & " " & Clients_LIST(p_ind).URL, "Diconnect")
        Try
            e.Close()
        Catch ex As Exception
            p_errlog("SocketException", "Close " & Clients_LIST(p_ind).Connected & " " & Clients_LIST(p_ind).URL, "Diconnect")
        End Try
        Try
            Dim list = ParseArrString(Clients_LIST(p_ind).TagsLst, ","c)
            For Each num In list
                'TagsList(Val(num.value)).SubsADD(id)
                TagsList(Val(num.value)).SubsDEL = p_ind
            Next
            Do While Clients_LIST(p_ind).tags_lst.Count > 0
                Clients_LIST(p_ind).tags_lst.RemoveAt(0)
            Loop
            Clients_LIST(p_ind).TEXTBOX_BUFF = ""
            Clients_LIST(p_ind).UserName = ""

            For i As Short = Clients_LIST.Count - 1 To 0 Step -1
                If Not Clients_LIST(i).IsActive Then
                    Clients_LIST.RemoveAt(i)
                Else
                    Exit For
                End If
            Next
        Catch ex As Exception
            Try
                p_errlog("SocketException", "Clients_LIST.RemoveAt(p_ind) " & Clients_LIST(p_ind).Connected & " " & Clients_LIST(p_ind).URL, "Disconnect")
            Catch ex3 As Exception
                p_errlog("SocketException", "Clients_LIST.RemoveAt(p_ind) " & "Clients_LIST(p_ind).Connected" & " " & "Clients_LIST(p_ind).URL", "Disconnect")
            End Try

        End Try
    End Sub
    
End Module

File: WpfWSS_CS/HTTP_ALL.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WpfWSS_CS
{
    using System;
    using System.Collections.Generic;
    using System.Net.Sockets;
    using System.Net;
    //using WSS.Client;
    //using Ma.netservice;
    using Microsoft.VisualBasic;

    public static class HttpAll
    {
        public static ulong v_http_conflict_count;
        public static ulong v_protectoin;
        public static bool chk_Deny_Enable = false;
        public static List<string> Deny_List = new List<string>();
        public static List<string> Users_List = new List<string>();
        public static double sim_1;
        public static double sim_2;
        public static double sim_3;
        public static bool send_data = false;
        public static bool simul_data = false;
        public static bool Wait_Bool = false;
        //public static MainWindow.Clients Clients_LIST = new Clients();
        public static long count_PAC, count_Byte, count_Byte_S, count_PAC_S, count_Err;
        public static long count_PAC_REC, count_Byte_REC, count_Byte_S_REC, count_PAC_S_REC, count_Err_REC;
        public static int Size_For_Send;
        public static short Wait_MSeconds;
        public static string Global_Data_For_Send;

        public static int F_ADD_CLIENT(string p_url, Socket p_Socket)
        {
            string[] HostArr = null;
            string HostName = "";

            if (!F_Get_Allow(p_url))
            {
                p_Socket.Close();
                return -1;
            }

            try
            {
                HostArr = p_url.Split(':');
                IPHostEntry myHost = Dns.GetHostEntry(HostArr[0]);
                HostName = myHost.HostName;

                if (!F_Get_Allow(HostName))
                {
                    p_Socket.Close();
                    return -1;
                }

                Users_List.Add(HostName);
            }
            catch (Exception )
            {
                HostName = "no DNSname";
            }

            for (int i = 0; i < MainWindow.Clients_LIST.Count; i++)
            {
                if (MainWindow.Clients_LIST[i].URL == p_url)
                {
                    return -1;
                }
                else if (!MainWindow.Clients_LIST[i].IsActive)
                {
                    MainWindow.Clients_LIST[i].URL = p_url;
                    MainWindow.Clients_LIST[i].ID = i;
                    MainWindow.Clients_LIST[i].Connected = DateTime.Now;
                    MainWindow.Clients_LIST[i].Stat = 0;
                    MainWindow.Clients_LIST[i].Socket = p_Socket;
                    MainWindow.Clients_LIST[i].IsActive = true;
                    MainWindow.Clients_LIST[i].MachineName = HostName;
                    MainWindow.Clients_LIST[i].ProcessReceive();
                    return i;
                }
            }

            MainWindow.Clients_LIST.Add(new Client(p_url, p_Socket));
            MainWindow.Clients_LIST[MainWindow.Clients_LIST.Count - 1].ID = MainWindow.Clients_LIST.Count - 1;
            MainWindow.Clients_LIST[MainWindow.Clients_LIST.Count - 1].MachineName = HostName;
            MainWindow.Clients_LIST[MainWindow.Clients_LIST.Count - 1].ProcessReceive();
                return MainWindow.Clients_LIST.Count - 1;
        }

        public static bool F_Get_Allow(string Str_Deny)
        {
            if (!chk_Deny_Enable)
            {
                return true;
            }

            bool v_ret = true;

            for (int j = 0; j < 2; j++)
            {
                switch (j)
                {
                    case 0:
                        foreach (string item in Deny_List)
                        {
                            if (Str_Deny.ToUpper().Contains(item.Substring(3)) && item.Substring(0, 3) == "DN=")
                            {
                                v_ret = false;
                            }
                        }
                        break;

                    case 1:
                        foreach (string item in Deny_List)
                        {
                            if (Str_Deny.ToUpper().Contains(item.Substring(3)) && item.Substring(0, 3) == "AL=")
                            {
                                v_ret = true;
                            }
                        }
                        break;
                }
            }

            return v_ret;
        }

        public static void SocketException(Socket e, short p_ind)
        {
            try
            {
                MainWindow.Clients_LIST[p_ind].IsActive = false;
                P_Errlog("SockExeption ", MainWindow.Clients_LIST[p_ind].URL.PadRight(23) + "|" + "Delete_Client".PadRight(30) + " | " +
                         MainWindow.Clients_LIST[p_ind].UserName.PadRight(23), "Diconnect");
            }
            catch (Exception)
            {
            }

            try
            {
                e.Close();
            }
            catch (Exception)
            {
                P_Errlog("SocketException", "Close " + MainWindow.Clients_LIST[p_ind].Connected + " " + MainWindow.Clients_LIST[p_ind].URL, "Diconnect");
            }

            try
            {
                string[] list = ParseArrString(MainWindow.Clients_LIST[p_ind].TagsLst, ',');
                foreach (var num in list)
                {
                    Dispatch.TagsList[Convert.ToInt32(num)].SubsDEL = p_ind;
                }

                while (MainWindow.Clients_LIST[p_ind].tags_lst.Count > 0)
                {
                    MainWindow.Clients_LIST[p_ind].tags_lst.RemoveAt(0);
                }

                MainWindow.Clients_LIST[p_ind].TEXTBOX_BUFF = "";
                MainWindow.Clients_LIST[p_ind].UserName = "";

                for (int i = MainWindow.Clients_LIST.Count - 1; i >= 0; i--)
                {
                    if (!MainWindow.Clients_LIST[i].IsActive)
                    {
                        MainWindow.Clients_LIST.RemoveAt(i);
                    }
                    else
                    {
                        break;
                    }
                }
            }
            catch (Exception)
            {
                try
                {
                    P_Errlog("SocketException", "Clients_LIST.RemoveAt(p_ind) " + MainWindow.Clients_LIST[p_ind].Connected + " " +
                                        MainWindow.Clients_LIST[p_ind].URL, "Disconnect");
                }
                catch (Exception)
                {
                    P_Errlog("SocketException", "Clients_LIST.RemoveAt(p_ind) " + "Clients_LIST(p_ind).Connected" + " " +
                                        "Clients_LIST(p_ind).URL", "Disconnect");
                }
            }
        }

        public static string[] ParseArrString(string input, char separator)
        {
            return input.Split(separator);
        }

        public static void P_Errlog(string p_point, string p_event, string p_msg)
        {
            try
            {
                int f_num = Convert.ToInt16(FreeFile());
                FileOpen(f_num, DateTime.Now.ToString("dd-MM-yyyy") + ".log", OpenMode.Append);
                WriteLine(f_num, DateTime.Now.ToString("dd.MM.yyyy HH:mm:ss") + " " + p_point + " | " + p_event + " | " + p_msg);
                FileClose(f_num);
            }
            catch (Exception)
            {
            }
        }

        public static int FreeFile()
        {
            // Реализуйте логику получения свободного номера файла.
            return 0;
        }

        public static void FileOpen(int fileNumber, string fileName, OpenMode mode)
        {
            // Реализуйте логику открытия файла с заданным номером.
        }

        public static void FileClose(int fileNumber)
        {
            // Реализуйте логику закрытия файла с заданным номером.
        }

        public static void WriteLine(int fileNumber, string line)
        {
            // Реализуйте логику записи строки в файл с заданным номером.
        }

    }

}
